ethexe/service/src/fast_sync.rs

```diff
@@ -98,81 +107,155 @@ impl EventData {
-                .expect("observer must fulfill database");
-            let previous_block_header = db
-                .block_header(previous_committed_block)
-                .expect("observer must fulfill database");
-            assert!(previous_block_header.height < latest_block_header.height);
+            let previous_committed_block =
+                Self::get_block_data(observer, db, previous_committed_block).await?;
+            debug_assert!(
+                previous_committed_block.header.height < latest_committed_block.header.height
+            );
         }
 
         Ok(Some(Self {
-            program_states,
-            program_code_ids,
-            needs_instrumentation_codes,
             latest_committed_batch,
             latest_committed_block,
             previous_committed_block,
         }))
     }
 }
 
+async fn net_fetch(
+    network: &mut NetworkService,
+    request: db_sync::Request,
+) -> Result<db_sync::Response> {
+    let request_id = network.db_sync().request(request);
+    loop {
+        let event = network
+            .next()
+            .await
+            .expect("network service stream is infinite");
+
+        if let NetworkEvent::DbResponse {
+            request_id: rid,
+            result,
+        } = event
+        {
+            debug_assert_eq!(rid, request_id, "unknown request id");
+            match result {
+                Ok(response) => break Ok(response),
+                Err((request, err)) => {
+                    log::warn!("Request {:?} failed: {err}. Retrying...", request.id());
+                    network.db_sync().retry(request);
+                    continue;
+                }
+            }
+        }
+    }
+}
+
+async fn collect_program_code_ids(
+    observer: &mut ObserverService,
+    network: &mut NetworkService,
+    latest_committed_block: H256,
+) -> Result<BTreeMap<ActorId, CodeId>> {
+    let router_query = observer.router_query();
+    let programs_count = router_query
+        .programs_count_at(latest_committed_block)
+        .await?;
+
+    let response = net_fetch(
+        network,
+        db_sync::Request::program_ids(latest_committed_block, programs_count),
+    )
+    .await?;
+
+    let program_code_ids = response.unwrap_program_ids();
+    Ok(program_code_ids)
+}
+
+async fn collect_code_ids(
+    observer: &mut ObserverService,
+    network: &mut NetworkService,
+    db: &Database,
+    latest_committed_block: H256,
+) -> Result<BTreeSet<CodeId>> {
+    let router_query = observer.router_query();
+    let codes_count = router_query
+        .validated_codes_count_at(latest_committed_block)
+        .await?;
+
+    let response = net_fetch(
+        network,
+        db_sync::Request::valid_codes(latest_committed_block, codes_count),
+    )
+    .await?;
+
+    let code_ids = response
+        .unwrap_valid_codes()
+        .into_iter()
+        .filter(|&code_id| db.code_valid(code_id).is_none())
```

Comment:
What if validation status exists but different from received?

Reply:
But you receives them from the network. That means that somebody share with use not valid codes also. At least a warning should be printed in that case